### PR TITLE
Homework

### PR DESCRIPTION
**Pull request by NeoBauer**
1. Created std::unique_ptr for engines in electric and petrol car;
2. Added speed to car interface, to check if we are moving or not,
3. Made function to to take away unique_ptr with engine "pullOutPetrolEngine()" when speed = 0;
4. Made getters to Electric and Petrol engines data "getElectricEngineData()", getPetrolEngineData(), they are returning raw ptr
5. Made accelerate() virtual function
6. Implemented accelerate() in ElectricCar, PetrolCar, and HybridCar to check if there are engines, and if a<0;
7. Accelerate() throw exception "invalid argument", when a<0 or there are no engines in car
8. Added "int speed" in Car.hpp
9. Added Exception checking if choosen gear is in given range (-1 to gears_)
10. Added few tests

### CODE REVIEW

File: ElectricCar.cpp [hunk 1]
```diff
@@ -1,16 +1,45 @@
...
+      std::cout << "Electric engine is pulled out\n";
+  return std::move(engine_);
+    }else throw std::runtime_error("This car is moving, you can't pull out engine\n");
+}
+
+void ElectricCar::restore(){
+    charge();
+}
+
+void ElectricCar::accelerate(int a) {
```
Comment (MaciejSalamonski) [inline ElectricCar.cpp:30]: very nice but i suggest you not to use variables like 'a, b, c'

**Review by MaciejSalamonski — APPROVED** [review]
Nice code! :~}

File: ElectricCar.cpp [hunk 2]
```diff
@@ -1,16 +1,45 @@
...
+  if(engine_ != nullptr)
+  return engine_.get();
+  else throw std::runtime_error("You can't get to data, because engine doesn't exists\n");
+}
+
+std::unique_ptr<ElectricEngine> ElectricCar::pullOutElectricEngine() {
+    if(speed == 0){
+      std::cout << "Electric engine is pulled out\n";
+  return std::move(engine_);
+    }else throw std::runtime_error("This car is moving, you can't pull out engine\n");
```
Comment (WBigaj) [inline ElectricCar.cpp:23]:
```suggestion
    if(speed == 0){
      std::cout << "Electric engine is pulled out\n";
      return std::move(engine_);
    }
    throw std::runtime_error("This car is moving, you can't pull out engine\n");
```

File: ElectricCar.cpp (hunk 1 above)
Comment (WBigaj) [reply ElectricCar.cpp:30]: +1

File: ElectricCar.cpp [hunk 3]
```diff
@@ -1,16 +1,45 @@
...
+    charge();
+}
+
+void ElectricCar::accelerate(int a) {
+  if (engine_) {
+    if (a >= 0) {
+      std::cout << "Accelerate to: " << a << "km\\h\n";
+      speed = a;
+    } else
+      throw std::invalid_argument("Invalid speed, can't be negative\n");
```
Comment (WBigaj) [inline ElectricCar.cpp:36]:
```suggestion
    }
    throw std::invalid_argument("Invalid speed, can't be negative\n");
```

File: HybridCar.cpp [hunk 4]
```diff
@@ -1,19 +1,23 @@
...
+void HybridCar::restore(){
+ElectricCar::restore();
+PetrolCar::restore();
+}
+
+void HybridCar::accelerate(int a){
+    if(PetrolCar::getPetrolEngineData() && ElectricCar::getElectricEngineData()) {
+        if(a > 0)
+        std::cout << "\nAccelerate to: " << a << "km\\h\n";
+        else throw std::invalid_argument("\nInvalid speed, can't be negative\n");
```
Comment (WBigaj) [inline HybridCar.cpp:21]:
```suggestion
        if(a > 0) {
          std::cout << "\nAccelerate to: " << a << "km\\h\n";
        }
        throw std::invalid_argument("\nInvalid speed, can't be negative\n");
```

File: HybridCar.cpp [hunk 5]
```diff
@@ -1,19 +1,23 @@
...
+    } else std::cout << "\nThis car doesn't have an engine!\n";
```
Comment (WBigaj) [inline HybridCar.cpp:22]:
```suggestion
    } else {
      std::cout << "\nThis car doesn't have an engine!\n";
    }
```

File: PetrolCar.cpp [hunk 6]
```diff
@@ -1,16 +1,33 @@
...
+std::unique_ptr<PetrolEngine> PetrolCar::pullOutPetrolEngine() {
+  return std::move(engine_);
+}
+
+void PetrolCar::accelerate(int a){
+    if(engine_){
+        if(a >= 0)
+        std::cout << "Accelerate to: " << a << "km\\h\n";
+        else throw std::invalid_argument("Invalid speed, can't be negative\n");
+    } else std::cout << "This car doesn't have an engine!\n";
```
Comment (WBigaj) [inline PetrolCar.cpp:32]:
```suggestion
    if(engine_) {
        if(a >= 0) {
           std::cout << "Accelerate to: " << a << "km\\h\n";
        } else {
           throw std::invalid_argument("Invalid speed, can't be negative\n");
        }
    } else {
        std::cout << "This car doesn't have an engine!\n";
    }
```

**Review by WBigaj — COMMENTED** [review]
Here is few suggestion how to fix code formatting. I recommend using clang-format.

**Review by Kasperice — APPROVED** [review]
I don't have any comments to the code. I see some inconsistent fomratting. Is the file 'temporary' necessary in this project?

File: HybridCar.cpp [hunk 7]
```diff
@@ -1,19 +1,26 @@
...
+
+void HybridCar::accelerate(int requiredSpeed) {
+  if (PetrolCar::getPetrolEngineData() &&
+      ElectricCar::getElectricEngineData()) {
+    if (requiredSpeed > 0) {
+      std::cout << "\nAccelerate to: " << requiredSpeed << "km\\h\n";
+    } else
+      throw std::invalid_argument("\nInvalid speed, can't be negative\n");
+  } else
+    std::cout << "\nThis car doesn't have an engine!\n";
```
Comment (ziobron) [inline HybridCar.cpp:25]:
```suggestion
    } else {
      throw std::invalid_argument("\nInvalid speed, can't be negative\n");
  } else {
    std::cout << "\nThis car doesn't have an engine!\n";
  }
```

File: PetrolCar.hpp [hunk 8]
```diff
@@ -1,16 +1,18 @@
...
-    void refuel();
+  PetrolCar(std::unique_ptr<PetrolEngine> engine);
+  ~PetrolCar();
 
-    PetrolEngine* engine_;
-};
+  void restore() override;
+  void accelerate(int a) override;
+  PetrolEngine *getPetrolEngineData() const;
+  std::unique_ptr<PetrolEngine> pullOutPetrolEngine();
```
Comment (ziobron) [inline PetrolCar.hpp:13]: where is putInPetrolEngine function? It is not symetrical with ElectricCar interface.

File: HybridCar.hpp [hunk 9]
```diff
@@ -1,19 +1,17 @@
...
-    void refuel();
+class HybridCar : public PetrolCar, public ElectricCar {
 
-    PetrolEngine* petrolEngine_;
-    ElectricEngine* electricEngine_;
+public:
+  HybridCar(std::unique_ptr<PetrolEngine> petrolEng,
+            std::unique_ptr<ElectricEngine> electricEng);
+  ~HybridCar();
+  PetrolEngine *getPetrolEngine() const;
```
Comment (ziobron) [inline HybridCar.hpp:14]: Why getPetrolEngine only? Why not getElectricEngine?
BTW I don't see an implementation of this function.

File: PetrolEngine.cpp [hunk 10]
```diff
@@ -2,19 +2,18 @@
...
-    // 0 is for NEUTRAL
+void PetrolEngine::changeGear(int gear) {
+  // TODO: Add checking if gear is between -1 and gears_
+  // -1 is for REAR
+  // 0 is for NEUTRAL
+  if (gear >= -1 && gear < gears_) {
     currentGear_ = gear;
-    std::cout << __FUNCTION__ << std::endl;
+  } else
+    throw WrongGearException(gear, this->gears_);
```
Comment (ziobron) [inline PetrolEngine.cpp:16]:
```suggestion
  } else {
    throw WrongGearException(gear, this->gears_);
  }
```

File: WrongGearException.cpp [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#include "WrongGearException.hpp"
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+
+const char *WrongGearException::what() const noexcept { return msg.c_str(); }
```
Comment (ziobron) [inline WrongGearException.cpp:6]: It's better not to override what() function, but pass a message to a proper constructor of a base class. You can use std::logic_error for example and you'll have it for free.

File: temporary [hunk 12]
```diff
@@ -0,0 +1,28 @@
+#include "WrongGearException.hpp"
```
Comment (ziobron) [inline temporary:1]: what is this file? I guess it shouldn't be commited.

File: tests.cpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+// Electric Car is best implemented, so I will check only electric Car
+
+SCENARIO("Checking if Electric functions works", "[ElectricCar]") {
+
+  GIVEN("ElectricCar, car raw_ptr and Shop") {
+    ElectricCar Tesla(std::make_unique<ElectricEngine>(600, 4000));
+    std::vector<std::unique_ptr<ElectricEngine>> OldEngineShop;
+
+    REQUIRE(Tesla.getElectricEngineData()->getBatteryCapacity() == 4000);
+    REQUIRE(Tesla.getElectricEngineData()->getPower() >= 600);
```
Comment (ziobron) [inline tests.cpp:16]: If you reuse some data it's better to assign it to a variable. It's a Demeter rule.

```suggestion
    auto electricEngine = Tesla.getElectricEngineData();
    REQUIRE(electricEngine->getBatteryCapacity() == 4000);
    REQUIRE(electricEngine->getPower() >= 600);
```

The same for all other places

File: tests.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+      OldEngineShop.push_back(Tesla.pullOutElectricEngine());
+      Tesla.putInElectricEngine(std::make_unique<ElectricEngine>(400, 3000));
+
+      THEN("Check if there are new parameters") {
+        REQUIRE(Tesla.getElectricEngineData()->getBatteryCapacity() == 3000);
+        REQUIRE(Tesla.getElectricEngineData()->getPower() >= 400);
+      }
+    }
+  }
+}
```
Comment (ziobron) [inline tests.cpp:53]: I can see only tests for an Electric Car. What about others?

**Review by ziobron — CHANGES_REQUESTED** [review]
Generally, good piece of code. Memory is managed well (unique_ptr) but interfaces are inconsistent. It looks that you have focused only on ElectricCar part. If you can make PetrolCar symmetrical it will be really good :)

File: WrongGearException.hpp [hunk 15]
```diff
@@ -1,13 +1,9 @@
-#include <exception>
+#include <stdexcept>
 #include <string>
-class WrongGearException : public std::exception {
+
+class WrongGearException : public std::logic_error {
 
 public:
+  using std::logic_error::logic_error;
```
Comment (ziobron) [inline WrongGearException.hpp:7]: nice :)

File: tests.cpp [hunk 16]
```diff
@@ -2,52 +2,203 @@
...
+    int speed = 50;
+    PetrolCar Golf(
+        std::make_unique<PetrolEngine>(power1, cubicCapacity1, gear_number1));
+    std::vector<std::unique_ptr<PetrolEngine>> OldEngineShop;
+
+    REQUIRE(Golf.getPetrolEngineData()->getCapacity() == cubicCapacity1);
+    REQUIRE(Golf.getPetrolEngineData()->getPower() >= power1);
+
+    WHEN("car is moving, and checking car data") {
+      int ActualGear = 3;
```
Comment (ziobron) [inline tests.cpp:84]: you mix UpperCamelCase with lowerCamelCase in variable names. Please use uniform way.

File: tests.cpp [hunk 17]
```diff
@@ -2,52 +2,203 @@
...
+      THEN("Cannot pull out engine") {
+        REQUIRE_THROWS_AS(Golf.pullOutPetrolEngine(), std::runtime_error);
+        REQUIRE(Golf.getPetrolEngineData()->getCapacity() == cubicCapacity1);
+        REQUIRE(Golf.getPetrolEngineData()->getPower() >= power1);
+        REQUIRE(OldEngineShop.size() == 0);
+        REQUIRE(Golf.getPetrolEngineData()->currentGear_ == ActualGear);
+      }
+    }
+    WHEN("car has stopped") {
+      Golf.accelerate(stopSpeed);
```
Comment (ziobron) [inline tests.cpp:96]: don't you use brake()?

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
7 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**